### ISSUE TITLE
Connection remain in INACTIVE state for prolonged period of time.

### ISSUE THREAD

**Issue by himanshu0791** [redis/lettuce#2930]
## Bug Report

I noticed that cached write connection from **getWriteConnection(slot)** from **PooledClusterConnectionProvider.java** remained in inactive state for indefinite period of time. This implementation is different from **getReadConnection(slot)** where we check if a connection is active before returning it to the caller.

#### Current Behavior
Connections remains in inactive state leading to **Currently not connected. Commands are rejected** for all requests based on config **rejectCommandsWhenInactive**.

<details>
In order to rule out any infra related issues where a netty channel never becomes active marking a connection as inactive, I tried Jedis client in parallel which do not run into any connection related issues.
This is primarily to see if there are variables outside of the client affecting it. I tried versions 5.3.7, 6.2.2 and they both observed same issue.

</details>

#### Possible Solution
I made some changes in the **getWriteConnection(slot)** similar to read path and that effectively fixed the issue. Essentially whenever a node connection is requested, I check for the status and return a new one instead if its marked as inative. I understand this is a bit inefficient in the sense that actual connections which may become active as also closed as part of this code fix.
below are the code snippets of the code I have modified.

`
        From PooledClusterConnectionProvider.java
        
        private CompletableFuture<StatefulRedisConnection<K, V>> getWriteConnection(int slot) {

        CompletableFuture<StatefulRedisConnection<K, V>> writer;// avoid races when reconfiguring partitions.
        synchronized (stateLock) {
            writer = writers[slot];
        }

        if (writer == null) {
        // writer instance is already completed. We can call get() without any blocking.
        if (writer == null || !writer.join().isOpen()) {
            synchronized (stateLock) {
                // reset slot connection so a new one can be saved. Connection if open, will be closed by
                // the underlying async connection provider.
                writers[slot] = null;
            }
            RedisClusterNode master = partitions.getMasterBySlot(slot);
            if (master == null) {
                clusterEventListener.onUncoveredSlot(slot);
                return Futures.failed(new PartitionSelectorException("Cannot determine a partition for slot " + slot + ".",
                        partitions.clone()));
            }

            // Use always host and port for slot-oriented operations. We don't want to get reconnected on a different
            // host because the nodeId can be handled by a different host.
            RedisURI uri = master.getUri();
            ConnectionKey key = new ConnectionKey(ConnectionIntent.WRITE, uri.getHost(), uri.getPort());

            ConnectionFuture<StatefulRedisConnection<K, V>> future = getConnectionAsync(key);

            return future.thenApply(connection -> {

                synchronized (stateLock) {
                    if (writers[slot] == null) {
                        writers[slot] = CompletableFuture.completedFuture(connection);
                    }
                }

                return connection;
            }).toCompletableFuture();
        }

        return writer;
    }
    
   protected ConnectionFuture<StatefulRedisConnection<K, V>> getConnectionAsync(ConnectionKey key) {

        ConnectionFuture<StatefulRedisConnection<K, V>> connectionFuture = connectionProvider.getConnection(key);
        ConnectionFuture<StatefulRedisConnection<K, V>> connectionFuture = connectionProvider.getConnection(key,
                StatefulConnection::isOpen, StatefulConnection::close);
        CompletableFuture<StatefulRedisConnection<K, V>> result = new CompletableFuture<>();

        connectionFuture.handle((connection, throwable) -> {

            if (throwable != null) {

                result.completeExceptionally(
                        RedisConnectionException.create(connectionFuture.getRemoteAddress(), Exceptions.bubble(throwable)));
            } else {
                result.complete(connection);
            }

            return null;
        });

        return ConnectionFuture.from(connectionFuture.getRemoteAddress(), result);
    }

    public CompletableFuture<StatefulRedisConnection<K, V>> getConnectionAsync(ConnectionIntent connectionIntent,
           String host, int port) {

        try {
            beforeGetConnection(connectionIntent, host, port);

            return connectionProvider.getConnection(new ConnectionKey(connectionIntent, host, port)).toCompletableFuture();
            return connectionProvider.getConnection(new ConnectionKey(connectionIntent, host, port),
                            StatefulConnection::isOpen, StatefulConnection::close)
                    .toCompletableFuture();
        } catch (RedisException e) {
            throw e;
        } catch (RuntimeException e) {
            throw new RedisException(e);
        }
    }
    
    
 From  AsyncConnectionProvider.java
 
     public F getConnection(K key, Predicate<T> isActive, Consumer<T> closeConnection) {
        return getSynchronizer(key, isActive, closeConnection).getConnection();
    }

    private Sync<K, T, F> getSynchronizer(K key, Predicate<T> isActive, Consumer<T> closeConnection) {

        if (closed) {
            throw new IllegalStateException("ConnectionProvider is already closed");
        }

        Sync<K, T, F> sync = connections.get(key);

        if (sync != null) {
            if (sync.isInProgress() || (sync.isComplete() && isActive.test(sync.connection))) {
                return sync;
            }
            // closing to avoid leak in case connection is deactivated and not closed.
            closeConnection.accept(sync.connection);
        }
        // reset key so a new connection can be cached.
        connections.remove(key);

        AtomicBoolean atomicBoolean = new AtomicBoolean();

        sync = connections.computeIfAbsent(key, connectionKey -> {

            Sync<K, T, F> createdSync = new Sync<>(key, connectionFactory.apply(key));

            if (closed) {
                createdSync.cancel();
            }

            return createdSync;
        });

        if (atomicBoolean.compareAndSet(false, true)) {

            sync.getConnection().whenComplete((c, t) -> {

                if (t != null) {
                    connections.remove(key);
                }
            });
        }

        return sync;
    }

   //Added a new method   
   private boolean isInProgress() {
       return PHASE.get(this) == PHASE_IN_PROGRESS;
   }  

    protected CompletableFuture<StatefulRedisConnection<K, V>> getConnection(RedisNodeDescription redisNodeDescription) {

        RedisURI uri = redisNodeDescription.getUri();

        return connectionProvider.getConnection(toConnectionKey(uri), StatefulConnection::isOpen, StatefulConnection::close).toCompletableFuture();
    }  
`

#### Additional context
We baked this fix for over a month on an industry standard code with QPS in order of millions without any issues. Logging did show us that there were multiple instances where this fix helped a rquest go through instead of failing with above mentioned exception.

I would like your thoughts on this and if this is something we can turn into a pull request.

**Comment by himanshu0791** [comment]
@mp911de Appreciate your comment on this.

**Comment by tishun** [comment]
@himanshu0791 are you still considering filing a PR?

I will try out your changes today and let you know if I have some concerns.

**Comment by tishun** [comment]
From a quick glance this seems like an omission from #503 where (a similar problem) was solved by checking the health of connections that are being returned. The `writers` portion seems to have been largely left unchanged since then.

> I understand this is a bit inefficient in the sense that actual connections which may become active as also closed as part of this code fix.

Not sure about this part myself, since ...

> ... noticed that some of the cached write connection from getWriteConnection(slot) from PooledClusterConnectionProvider.java remain in inactive state for indefinite period of time. 

`indefinite period of time` does not seem like a reasonable use case to wait for a write connection.

Overall I think your suggestion makes sense, perhaps with the exception that we might not want to block here

```java
        if (writer == null || !writer.join().isOpen()) {
```

I'd try to incorporate this in the future chain somehow.

**Comment by PreAgile** [comment]
Hi, I'd like to work on this issue.

  From the discussion so far, it looks like the write-side cached connection path in
  `PooledClusterConnectionProvider` is missing the kind of connection health check that was added for the
  similar case in #503. I’m planning to follow that direction, but keep the implementation non-blocking
  and avoid introducing a `.join()` in this path, as suggested by @tishun.

  I’ll start by comparing the current read/write connection paths and then add a focused test for a
  cached write connection that is no longer usable before proposing the fix.

**Comment by tishun** [comment]
@PreAgile please feel free to work on that if you have the cycles. Send a PR when you are ready.

We do not typically assign the issues to anybody, so just send out a message so other contributors know you are working on it.

**Comment by PreAgile** [comment]
Took a swing at this in #3763. The cached writer is now health-checked with non-blocking primitives (`isDone` / `getNow(null)` / `isOpen`) and, when the channel is closed, the slot is evicted under `stateLock` together with the matching `AsyncConnectionProvider` entry. The provider entry is located by identity match through `forEach` so a topology change that has already remapped the slot cannot lead to closing the fresh connection. In-flight futures are returned as-is to avoid preempting concurrent connection attempts.

Mirrors the read-path recovery from #503; `AsyncConnectionProvider.close(K)` is reused so the underlying close is `closeAsync`, not blocking.

@tishun happy to iterate if you would like a different shape (for example moving the health check into `AsyncConnectionProvider` itself as a follow-up).

**Comment by a-TODO-rov** [comment]
What we basically have here is cache invalidation issue.
I think we should go in the direction of event-driven cache invalidation. I will place a design proposal below:

## 1. Problem

`PooledClusterConnectionProvider` maintains two slot-indexed caches — `writers[16384]` for write connections and `readers[16384][]` for read candidates. These caches are **write-once**: once a slot is populated with a connection, it is never re-validated. The only mechanism that clears them is `resetFastConnectionCache()`, triggered by topology refresh or provider shutdown.

When a cached connection's channel dies (network partition, node crash, idle timeout), the cache continues returning the dead connection to callers.

## 2. Root Cause

The slot caches have **no awareness of connection lifecycle events**. The system already knows when a connection dies — Netty fires `channelInactive`, which propagates through `ConnectionEventTrigger` and `DefaultEndpoint` to the connection's `deactivated()` callback. But none of these events reach the provider that cached the connection.

There is no back-reference from a connection to its provider. The relationship is one-directional: provider → connection.

## 3. Proposed Solution

### 3.1 Approach: Event-driven eviction

Subscribe to connection lifecycle events using the existing `RedisConnectionStateListener` infrastructure. When a connection reports disconnection, scan the slot caches and clear all entries pointing to that connection by identity.

### 3.2 Hook point

`NodeConnectionPostProcessor.apply()` already runs post-creation setup on every connection created by the provider (sets autoFlush, registers push listeners). The disconnection listener is registered here — same pattern, same location.

### 3.3 Eviction logic

On disconnection callback: acquire `stateLock`, iterate `writers[]` and `readers[][]`, null out any slot whose completed future resolves to the disconnected connection (identity match). Release lock.

**Critical detail — identity match, not `isOpen()`:** The Netty pipeline ordering means `onRedisDisconnected` fires *before* `deactivated()` sets `active = false`. At callback time, `isOpen()` still returns `true`. The disconnection event itself is the signal — identity match against the connection reference is the correct check.

### 3.4 What does NOT change

- `getWriteConnection()` — no modification needed. The existing code is correct *if* stale entries are evicted externally. The bug was never in the access path.
- `AsyncConnectionProvider` — physical connections remain managed by `Sync` self-removal and watchdog reconnection. Only the slot fast-path cache is affected.
- `RedisChannelHandler`, `ConnectionFacade`, `DefaultEndpoint` — no changes.
- Public API — no new interfaces, no new configuration.

## 4. Open Questions
1. **Scan Cost** Eviction scans all 16,384 writer slots and all 16,384 reader slots per disconnection event. This is the same cost profile as `resetFastConnectionCache()`  — already accepted in the codebase. **Possible enchantment** Reverse index:
 - Maintain an `IdentityHashMap<Connection, BitSet>` mapping each connection to its set of cached writer slots.
- Populate on slot caching (in `getWriteConnection` `thenApply`). Clear on `resetFastConnectionCache()`.
- On eviction: look up the `BitSet`, iterate only the relevant slots. 
2. **Proactive re-population on reconnect?** When the watchdog reconnects, `onRedisConnected` fires. We could re-populate slots proactively. However, this requires the reverse lookup (connection → slots), and lazy population on first access is simple and correct. **Recommendation: Not in scope.**
3. **Physical connection eviction from `AsyncConnectionProvider`?** After slot eviction, the stale `Sync` entry remains in the physical connection pool. If watchdog reconnects, the connection self-heals. If the node is truly dead, `Sync` self-removes via its failure callback. Proactively closing the physical connection would *prevent* watchdog recovery on transient failures. **Recommendation: No change.**
4. **Read path fallback.** The `selectedReaderCandidates[0]` dead-connection fallback becomes very narrow with proactive eviction (only fires on a snapshot race). Should it be changed to force re-selection? **Recommendation: Yes — defensive improvement, low risk.**

@tishun what do you think ?

**Comment by tishun** [comment]
Generally I agree with @a-TODO-rov that this piece of code (the whole `PooledClusterConnectionProvider`) could do with a lot of improvement. I do feel that this would require significantly more effort than the currently proposed solution, if we want to make it right. It would also require modifying the straight path, increasing the possibility of introducing regressions.

In the same time the solution in #3763 seems to follow the current patterns applied for the readers - lazy checks for stale connections; and is generally a lightweight solution to the existing problem.

My instinct is to go with the initial implementation by @PreAgile unless we can dedicate a more focused effort.

**Comment by PreAgile** [comment]
Thanks for laying out the design in such detail.

I agree the event-driven eviction is the more complete fix; it does not depend on the next access to notice the stale slot, and it covers the writer and reader caches in one place. My only hesitation about folding it into this PR is the point @tishun raised: doing it properly touches the straight path and widens the regression surface, while the current patch stays on the read-path pattern that already exists in the codebase.

So that the design is not lost, I am happy to open a follow-up issue capturing your proposal (listener registration in `NodeConnectionPostProcessor`, identity-match eviction under `stateLock`, and the reverse-index option for scan cost) and reference it as the longer-term direction. That keeps this PR as the lightweight fix for the immediate INACTIVE-connection case, with the fuller rework tracked separately.

Would that split work for you, or would you rather take on the rework yourself and hold #3763 until then?
